### DISCUSSION TITLE
How to make post request in controller?

### DISCUSSION THREAD

**Discussion by rayacode** [playframework/playframework#12791] [Q&A]
I'm implementing integration for a payment gateway, gateway docs shows a form with various hidden fields to be post submit to the gateway url, i don't want let those fields be available even as hidden in form rather i want to make a simple form with one text field for amount to be paid then in my controller i emulate the form post submit of gateway and user when hits the pay button posts happen and goes to processor, basically the question is how do emulate the form submit in controller and not with wsclient since i want user goes into gateway url, also i tried wsclient post but it return type not castable to CompleteStage<Result>

**Comment by mkurz** [comment]
> ...and not with wsclient since i want user goes into gateway url,

Sorry, I don't follow here, what do you mean by "_i want user goes into gateway url_"?

**Reply by rayacode** [reply]
i mean payment gateway url( of any payment processor provider like paypal, payeer) scenario i want to make is like this:
Twirl:
simple form with one input and one button, input value is the amount user want to deposit,
then in controller method,
i get the input value, and among with other parameters post it to certain URL.
lets have ex:
capitalist payment processor docs says:
 Payment Request Form
This HTML form is the key one. With its help you can send a customer for a payment. It provides the
basic parameters required for this purpose. As you can see from the example, you can pass the query parameter
lang=en or ru to select the interface language.
The method="get” is also available, so you can pass these parameters in a query string.
Request Example
`
form name="payment" method="post" action="https://capitalist.net/merchant/pay?lang=en"
accept-charset="UTF-8"><input type="hidden" name="merchantid" value="70" /
input type="hidden" name="number" value="123981" /
input type="hidden" name="amount" value="15.23" /
input type="hidden" name="currency" value="USD" /
input type="hidden" name="description" value="My description" /
input type="hidden" name="sign" value="5ae3dbfdbac751f44355dedf5db59d02" /
/form`

look here capitalist says have a form like this, but rather i want to have a form with one field and keep these hidden parameters in back-end not as hidden input fields in a template and post the form in back-end! when user hits the pay button post must happen by controller and it must then redirected to the capitalist payment page to process the payment

**Comment by mjackson-mim** [comment]
> i tried wsclient post but it return type not castable to CompleteStage

You should be able to get the `CompletionStage` by doing this, assuming you're in Java:

```java
CompletionStage<WSResponse> stream = wsClient.url("https://www.paymentservice.com").setMethod("POST").stream();
```

JavaDoc:

- [`WSClient.url()`](https://www.playframework.com/documentation/2.8.x/api/java/play/libs/ws/WSClient.html#url-java.lang.String-)
-  [`WSRequest.stream()`](https://www.playframework.com/documentation/2.8.x/api/java/play/libs/ws/WSRequest.html#stream--)

---

Hopefully I understand your question correctly, if not, feel free to correct me. You want the steps to be

1. Customer gets a form with the amount they owe, and a button to submit the form
2. Your website gets the POST request from the form
3. Your web app sends its own POST request (containing extra information) to the payment service, and receives a response
4. You then redirect the customer to a page on the payment service's website


If you're OK with the customer leaving your website, how you send the request doesn't matter, just return a 303 redirect response to the customer with the URL to the page you want to display. If you don't like `WSClient`, there should be plenty of open source libraries you can pull from with a search online that will happily send POST requests.

**Reply by mkurz** [reply]
Adding a link the the docs here as well: https://www.playframework.com/documentation/3.0.x/JavaWS
